### ISSUE TITLE
no-dupe-else-if, no-duplicate-case and no-unneeded-ternary rules are inconsistent.

### ISSUE THREAD

**Issue by ota-meshi** [eslint/eslint#13466]
**Tell us about your environment**

* **ESLint Version:** 7.4.0
* **Node Version:** 12.16.1
* **npm Version:** 6.13.4

**What parser (default, Babel-ESLint, etc.) are you using?**

Default parser.

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "rules": {
    "no-dupe-else-if": "error",
    "no-duplicate-case": "error",
    "no-unneeded-ternary": ["error", {"defaultAssignment": false }]
  }
}

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
if (a.b) {}
else if (a.b) {} // reported by no-dupe-else-if.
else if (a.     b) {} // reported by no-dupe-else-if.

switch (a) {
  case a.b:
  case a.b: // reported by no-duplicate-case.
  case a.      b: // Not reported
}

var foo = a.b ? a.b : bar; // Not reported
```

```bash
eslint .
```

**What did you expect to happen?**

I expected the no-duplicate-case and no-unneeded-ternary rules to report errors as well as the no-dupe-else-if rule.

**What actually happened? Please include the actual, raw output from ESLint.**

```bash
  3:10  error  This branch can never execute. Its condition is a duplicate or covered by previous conditions in the if-else-if chain  no-dupe-else-if
  4:10  error  This branch can never execute. Its condition is a duplicate or covered by previous conditions in the if-else-if chain  no-dupe-else-if
  8:3   error  Duplicate case label                                                                                                   no-duplicate-case
```

The no-duplicate-case and no-unneeded-ternary rules do not report errors with some patterns.  
You can check it with [DEMO].

This issue is similar to issue #11852, but I'm reporting inconsistency. If the getter has side effects, the `no-dupe-else-if` rule should not report property access (by default).

**Are you willing to submit a pull request to fix this bug?**

I can open a pull request if we can decide which behavior is correct.

[DEMO]: https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50XG4gIG5vLWR1cGUtZWxzZS1pZjogZXJyb3IsXG4gIG5vLWR1cGxpY2F0ZS1jYXNlOiBlcnJvcixcbiAgbm8tdW5uZWVkZWQtdGVybmFyeTogW2Vycm9yLCB7ZGVmYXVsdEFzc2lnbm1lbnQ6IGZhbHNlIH1dLFxuICAqL1xuXG5pZiAoYS5iKSB7fVxuZWxzZSBpZiAoYS5iKSB7fSAvLyByZXBvcnRlZCBieSBuby1kdXBlLWVsc2UtaWYuXG5lbHNlIGlmIChhLiAgICAgYikge30gLy8gcmVwb3J0ZWQgYnkgbm8tZHVwZS1lbHNlLWlmLlxuXG5zd2l0Y2ggKGEpIHtcbiAgY2FzZSBhLmI6XG4gIGNhc2UgYS5iOiAvLyByZXBvcnRlZCBieSBuby1kdXBsaWNhdGUtY2FzZS5cbiAgY2FzZSBhLiAgICAgIGI6IC8vIE5vdCByZXBvcnRlZFxufVxuXG52YXIgZm9vID0gYS5iID8gYS5iIDogYmFyOyAvLyBOb3QgcmVwb3J0ZWRcbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=

**Comment by mdjermanovic** [comment]
I don't think this is a bug, it's just different rules having different logic on how to compare two expressions, and one way might not be necessarily suitable for all rules.

That said, I think it makes sense to evaluate comparing tokens in `no-duplicate-case` and `no-unneeded-ternary` instead of the actual behavior, as enhancements to these rules.

As for `no-dupe-else-if`, I personally think it's fine as is. This rule would be pretty limited if it would compare only identifiers, and possible side effects are covered by the [When Not To Use It](https://eslint.org/docs/rules/no-dupe-else-if#when-not-to-use-it) section.

**Comment by yeonjuan** [comment]
I think `no-duplicate-case` should be smarter. For now, it just compares the node's text.

**Comment by ota-meshi** [comment]
Thank you for your reply!

> I don't think this is a bug

I did not intend to insist that this is a bug. I'm sorry if hurt your feelings.

> That said, I think it makes sense to evaluate comparing tokens in `no-duplicate-case` and `no-unneeded-ternary` instead of the actual behavior, as enhancements to these rules.

I think it's good, if you can change the `no-duplicate-case` and `no-unneeded-ternary` rules, user will be able to detect more problems. 
Can I send pull requests to change the `no-duplicate-case` rule and the `no-unneeded-ternary` rule to compare tokens?
Is there anything else I need to do?

**Comment by mdjermanovic** [comment]
> I think it's good, if you can change the `no-duplicate-case` and `no-unneeded-ternary` rules, user will be able to detect more problems.
> Can I send pull requests to change the `no-duplicate-case` rule and the `no-unneeded-ternary` rule to compare tokens?
> Is there anything else I need to do?

I personally support both changes :+1: , but since these are enhancements and thus require [consensus](https://eslint.org/docs/maintainer-guide/issues#accepting-issues), it might be better to create rule change request issues first. We should also evaluate if some of the changes have to be behind an additional option, or maybe need to wait for the next major version.

**Comment by ota-meshi** [comment]
@mdjermanovic Thank you for your advice! I have opened two new issues.  
I close this issue.
